src/include/duckdb/common/types/date_lookup_cache.hpp

```diff
@@ -0,0 +1,65 @@
+//===----------------------------------------------------------------------===//
+//                         DuckDB
+//
+// duckdb/common/types/date_lookup_cache.hpp
+//
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "duckdb/common/types/date.hpp"
+#include "duckdb/common/types/timestamp.hpp"
+#include "duckdb/common/types/value.hpp"
+#include "duckdb/common/types/validity_mask.hpp"
+
+namespace duckdb {
+struct ValidityMask;
+
+template <class OP>
+class DateLookupCache {
+public:
+	using CACHE_TYPE = uint16_t;
+	constexpr static int32_t CACHE_MIN_DATE = 0;     // 1970-01-01
+	constexpr static int32_t CACHE_MAX_DATE = 29584; // 2050-12-31
+
+public:
+	DateLookupCache() {
+		BuildCache();
+	}
+
+	//! Extracts the component, or sets the validity mask to NULL if the date is infinite
+	int64_t ExtractElement(date_t date, ValidityMask &mask, idx_t idx) const {
```

Comment:
Probably not important but there are two date parts that are doubles (Julian Day and Epoch) so this would need to be parameterized if we decided to optimize them.